File: backend/app/services/platform.py
from app.database import supabase
from app.services.crypto import encrypt_token, decrypt_token
from functools import lru_cache
import time

_cache: dict = {}
_cache_ttl = 300  # 5 minutos


def get_setting(key: str) -> str:
    now = time.time()
    if key in _cache and now - _cache[key]["ts"] < _cache_ttl:
        return _cache[key]["value"]

    result = supabase.table("platform_settings").select("value_encrypted").eq("key", key).single().execute()
    raw = result.data["value_encrypted"] if result.data else ""
    value = decrypt_token(raw) if raw else ""
    _cache[key] = {"value": value, "ts": now}
    return value


def set_setting(key: str, value: str):
    encrypted = encrypt_token(value) if value else ""
    supabase.table("platform_settings").upsert({
        "key": key,
        "value_encrypted": encrypted,
        "updated_at": "now()",
    }, on_conflict="key").execute()
    _cache.pop(key, None)


def get_anthropic_key() -> str:
    return get_setting("anthropic_api_key")


def get_meta_credentials() -> dict:
    return {
        "app_id": get_setting("meta_app_id"),
        "app_secret": get_setting("meta_app_secret"),
    }

File: frontend/lib/server/platform.ts
import { createServiceClient } from "./supabase"
import { decrypt, encrypt } from "./crypto"

const cache = new Map<string, { value: string; ts: number }>()
const TTL = 300_000

// Rows written before encryption was wired up here are stored as raw plaintext
// (no iv:tag:ciphertext structure). Decrypt when possible, fall back otherwise
// so already-configured settings don't break while they get re-saved.
function safeDecrypt(raw: string): string {
  if (!raw) return raw
  if (raw.startsWith("plain:")) return decrypt(raw)
  const parts = raw.split(":")
  if (parts.length === 3 && parts.every(p => /^[0-9a-f]+$/i.test(p))) {
    try { return decrypt(raw) } catch { /* fall through to legacy plaintext */ }
  }
  return raw
}

async function getSetting(key: string): Promise<string> {
  const cached = cache.get(key)
  if (cached && Date.now() - cached.ts < TTL) return cached.value

  const supabase = createServiceClient()
  const { data } = await supabase
    .from("platform_settings")
    .select("value_encrypted")
    .eq("key", key)
    .single()

  const value = safeDecrypt(data?.value_encrypted ?? "")
  cache.set(key, { value, ts: Date.now() })
  return value
}

export async function setSetting(key: string, value: string) {
  const supabase = createServiceClient()
  const { error } = await supabase
    .from("platform_settings")
    .update({ value_encrypted: encrypt(value), updated_at: new Date().toISOString() })
    .eq("key", key)
  if (error) throw new Error(error.message)
  cache.delete(key)
}

export const getAnthropicKey        = () => getSetting("anthropic_api_key")
export const getMetaAppId           = () => getSetting("meta_app_id")
export const getMetaAppSecret       = () => getSetting("meta_app_secret")
export const getGoogleClientId      = () => getSetting("google_client_id")
export const getGoogleClientSecret  = () => getSetting("google_client_secret")
export const getGoogleDeveloperToken = () => getSetting("google_developer_token")
export { getSetting }
